### ISSUE TITLE
Training with losses that adapt model silently adapts model for other losses

### ISSUE THREAD

**Issue by Andrewwango** [deepinv/deepinv#875]
Currently it's not trivial how to train with 2 losses where one loss adapts the model (e.g. splitting loss) but the other loss uses the whole measurement, like `loss = splitting + |Af(y) - y|^2`. Currently, when one does `loss = [Splitting, MC]`, the MC loss will silently use as `x_net` `f(y1)` rather than `f(y)`.

The only current way to do is to modify your second loss to do `x_net_orig = model.model(y); return metric(physics(x_net_orig),y)`. However this is hacky.

Note that losses generally should be calculated using the same reconstruction arising from the same model. However, some users will want to do as above, and expect that non-adapting losses use `x_net=f(y)`.

We should add a warning when this happens:

in trainer: `if len(losses) > 1 and len({isinstance(loss, SplittingLoss) for loss in losses}) > 1: warn(all losses will receive an x_net that is the output of a {model.__class__.__name__}, rather than the original model. Make sure this is intended.)` This warning should arise either when losses are instantiated or during the adapt_model step in the trainer.

The issue is that SplittingLoss is not the only one adapting the model. The best way would be either to do a base loss for these or a base adapted model for SplittingModel, R2RModel etc..

**Comment by jscanvic** [comment]
Agreed - thanks for raising this point!

What do you think of using duck typing here? It's a bit unfortunate we made the choice of making the base class `Loss` expose a dummy `adapt_model` method, otherwise we could simply test if the loss instance/its class has an attribute/method `adapt_model`. That way, we would simply test if there's more than one training loss with at least one of them adapting the model. As it turns out, we can still do that but it's going to be a bit more cumbersome. Instead of checking if the method exists (it should always exist with our current conventions), we can test if it's different from the base one.

(I believe the whole "model adaptation" is not the way forward and I hope we'll move away from it at some point so I'd prefer if we avoided to give it a more central place in the code base / type hierarchy)

Proof of concept:

```python
# poc.py
import deepinv as dinv

losses = [
    dinv.loss.SplittingLoss(split_ratio=0.25),
    dinv.loss.MCLoss()
]

for loss in losses:
    model_adapting_loss = getattr(type(loss), "adapt_model", dinv.loss.Loss.adapt_model) is not dinv.loss.Loss.adapt_model
    print(f"{type(loss).__name__} {"doesn't adapt" if not model_adapting_loss else "adapts"} the model")

```

```
$ python poc.py
SplittingLoss adapts the model
MCLoss doesn't adapt the model
```
